Unicode/ustring.c

```diff
@@ -562,20 +563,27 @@ void utf82u_strcat(unichar_t *to,const char *from) {
     utf82u_strcpy(to+u_strlen(to),from);
 }
 
-char *u2utf8_strcpy(char *utf8buf,const unichar_t *ubuf) {
+char *u2utf8_strncpy(char *utf8buf,const unichar_t *ubuf,int len) {
 /* Copy unichar string 'ubuf' into utf8 buffer string 'utf8buf' */
     char *pt = utf8buf;
 
+    assert(utf8buf != NULL);
+
     if ( ubuf!=NULL ) {
-	while ( *ubuf && (pt=utf8_idpb(pt,*ubuf++,0)) );
-	if ( pt ) {
-	    *pt = '\0';
-	    return( utf8buf );
-	}
+        while ( *ubuf && (--len != 0) ) {
+            pt=utf8_idpb(pt,*ubuf++,0);
```

Comment:
I didn't think about the loop restructuring but it does improve the code given the assert. The original author probably put things in that form thinking it protected against `pt==NULL` but it doesn't unless `utf8_idpb` does. 

Anyway this looks good to me. (I assume it passes our unit tests?)

Reply:
Yup, it does. I'm going to merge exactly this, but I'm going to force push my branch to consolidate the last three commits into one commit. That way we have separate commits for the Git history of (1) removing the second UTF-8 decoder implementation and (2) adding the new function. So if you see new commits, they aren't really new.